python/opae.admin/opae/admin/sysfs.py

```diff
@@ -0,0 +1,282 @@
+            yield sysfs_node(p)
+
+    @contextmanager
+    def _open(self, mode):
+        try:
+            with open(self._sysfs_path, mode) as fd:
+                yield fd
+        except IOError as ioerr:
+            self.log.exception(ioerr)
+            raise
+
+    @property
+    def value(self):
+        with self._open('r') as fd:
+            return fd.read().strip()
+
+    @value.setter
+    def value(self, val):
+        if DRY_RUN:
+            print('echo {} > {}'.format(val, self._sysfs_path))
+        else:
+            with self._open('w') as fd:
+                fd.write(val)
+
+
+    @property
+    def sysfs_path(self):
+        return self._sysfs_path
+
+
+class pci_node(sysfs_node):
+    PCI_BUS_SYSFS = '/sys/bus/pci/devices'
+    """pci_node is a class used to encapsulate a node on the pci bus
+       that can be found in /sys/bus/pci/devices and can have a parent
+       node and one or more children nodes
+       This can be used to represent a PCIe tree or subtree"""
+    def __init__(self, pci_address, parent=None, **kwargs):
+        """__init__ initialize a pci_node object
+
+        :param pci_address: The pci address of the node using following format
+                            [segment:]bus:device.function
+        :param parent(pci_node): Another pci_node object that is the parent of
+                                 this node in the PCIe tree
+        """
+        node_path = os.path.join(self.PCI_BUS_SYSFS,
+                                 pci_address['pci_address'])
+        super(pci_node, self).__init__(node_path)
+        self._pci_address = pci_address
+        self._parent = parent
+        self._children = []
+        self._aer_cmd1 = 'setpci -s {} ECAP_AER+0x08.L'.format(
+            pci_address['pci_address'])
+        self._aer_cmd2 = 'setpci -s {} ECAP_AER+0x14.L'.format(
+            pci_address['pci_address'])
+
+    def __str__(self):
+        return '[pci_address({}), pci_id(0x{:04x}, 0x{:04x})]'.format(
+            self.pci_address, *self.pci_id)
+
+    def __repr__(self):
+        return str(self)
+
+    def _find_children(self):
+        children = []
+        for f in os.listdir(self.sysfs_path):
+            if not f.startswith(self.pci_address):
+                m = PCI_ADDRESS_RE.match(f)
+                if m:
+                    children.append(pci_node(m.groupdict(), self))
+        return children
+
+    def tree(self, level=0):
+        text = '{}{}\n'.format(' ' * level*4, self)
+        for n in self.children:
+            text += n.tree(level+1)
+        return text
+
+    @property
+    def root(self):
+        if self.parent is None:
+            return self
+        return self.parent.root
+
+    @property
+    def pci_address(self):
+        """pci_address get the pci address of the node"""
+        return self._pci_address['pci_address']
+
+    @property
+    def bdf(self):
+        """pci_address get the pci address of the node"""
```

Comment:
Comment seems to be a copy/paste of line 139.  Should make it specific to bdf() I think.